AsyncDisplayKit/Layout/ASLayoutOptions.mm

```diff
@@ -0,0 +1,252 @@
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#import "ASLayoutOptions.h"
+
+#import <AsyncDisplayKit/ASAssert.h>
+#import <AsyncDisplayKit/ASThread.h>
+#import <AsyncDisplayKit/ASTextNode.h>
+#import "ASInternalHelpers.h"
+
+@interface ASLayoutOptions()
+{
+  ASDN::RecursiveMutex _propertyLock;
+}
+@end
+
+@implementation ASLayoutOptions
+
+@synthesize spacingBefore = _spacingBefore;
+@synthesize spacingAfter = _spacingAfter;
+@synthesize flexGrow = _flexGrow;
+@synthesize flexShrink = _flexShrink;
+@synthesize flexBasis = _flexBasis;
+@synthesize alignSelf = _alignSelf;
+
+@synthesize ascender = _ascender;
+@synthesize descender = _descender;
+
+@synthesize sizeRange = _sizeRange;
+@synthesize layoutPosition = _layoutPosition;
+
+static Class gDefaultLayoutOptionsClass = nil;
++ (void)setDefaultLayoutOptionsClass:(Class)defaultLayoutOptionsClass
+{
+  gDefaultLayoutOptionsClass = defaultLayoutOptionsClass;
+}
+
++ (Class)defaultLayoutOptionsClass
+{
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    if (gDefaultLayoutOptionsClass == nil) {
+      // If someone is asking for this and it hasn't been customized yet, use the default.
+      gDefaultLayoutOptionsClass = [ASLayoutOptions class];
+    }
+  });
+  return gDefaultLayoutOptionsClass;
+}
+
+- (instancetype)init
+{
+  return [self initWithLayoutable:nil];
+}
+
+- (instancetype)initWithLayoutable:(id<ASLayoutable>)layoutable;
+{
+  self = [super init];
+  if (self) {
+    [self setupDefaults];
+    [self setValuesFromLayoutable:layoutable];
+  }
+  return self;
+}
+
+#pragma mark - NSCopying
+- (id)copyWithZone:(NSZone *)zone
+{
+  ASLayoutOptions *copy = [[[self class] alloc] init];
+  [self copyIntoOptions:copy];
+  return copy;
+}
+
+- (void)copyIntoOptions:(ASLayoutOptions *)copy
+{
+  ASDN::MutexLocker l(_propertyLock);
+  copy.flexBasis = self.flexBasis;
+  copy.spacingAfter = self.spacingAfter;
+  copy.spacingBefore = self.spacingBefore;
+  copy.flexGrow = self.flexGrow;
+  copy.flexShrink = self.flexShrink;
+  
+  copy.ascender = self.ascender;
+  copy.descender = self.descender;
+  
+  copy.sizeRange = self.sizeRange;
+  copy.layoutPosition = self.layoutPosition;
+}
+
+
+#pragma mark - Defaults
+- (void)setupDefaults
+{
+  self.flexBasis = ASRelativeDimensionUnconstrained;
+  self.spacingBefore = 0;
+  self.spacingAfter = 0;
+  self.flexGrow = NO;
+  self.flexShrink = NO;
+  self.alignSelf = ASStackLayoutAlignSelfAuto;
+
+  self.ascender = 0;
+  self.descender = 0;
+  
+  self.sizeRange = ASRelativeSizeRangeMake(ASRelativeSizeMakeWithCGSize(CGSizeZero), ASRelativeSizeMakeWithCGSize(CGSizeZero));
+  self.layoutPosition = CGPointZero;
+}
+
+// Do this here instead of in Node/Spec subclasses so that custom specs can set default values
```

Comment:
Could you refine this comment?  I don't understand it...if I have a custom spec, how would that help me set default values, with this method implemented in the Options (not spec)?